### ISSUE TITLE
Some file get corrupted by zero filled with relatime=on compression=zstd dedup=verify trim

### ISSUE THREAD

**Issue by logoerthiner1** [openzfs/zfs#13784]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Fedora
Distribution Version	| 36
Kernel Version	| 5.18.17-200.fc36.x86_64
Architecture	| x86_64
OpenZFS Version	| 2.1.5-1

Additional: Under Xen 4.14.5, with balloon, zfs partition is a LVM volume in a thin pool. Volume is 100G; VM has 20G RAM and 6 cpu cores. Docker version 20.10.17, build 100c701

### Describe the problem you're observing

I was playing with [docker on ZFS](https://docs.docker.com/storage/storagedriver/zfs-driver/) and tinkering with zfs options.

When I have no additional options (`zpool create -f zpool-docker -m /var/lib/docker /dev/xvdi`), docker pull & build are all good.

I have tried `-O relatime=on -O compression=zstd -O dedup=verify` and then building the same Dockerfile, and at one step it pops out some strange error:

```
Error loading shared library libpython3.10.so.1.0: Exec format error (needed by /usr/bin/python3)
Error relocating /usr/bin/python3: Py_BytesMain: symbol not found
```

Further inspecting in the container and I find out a scary fact:
```
$ hexdump -C /usr/lib/libpython3.10.so.1.0
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
002b8010
$ hexdump -C /usr/bin/find
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0003bc60
```

It seems that the file has its content being all zeros.

The files are just installed with package manager.

There may be other file "infected" as well - I did not check out.

`zpool scrub` does not think that there is a problem.

```
$ sudo zpool status -v zpool-docker
  pool: zpool-docker
 state: ONLINE
  scan: scrub repaired 0B in 00:06:57 with 0 errors on Fri Aug 19 17:57:42 2022
config:

	NAME          STATE     READ WRITE CKSUM
	zpool-docker  ONLINE       0     0     0
	  xvdi        ONLINE       0     0     0

errors: No known data errors
```

Things probably related:
1. At that time I am concurrently building a dockerfile and pulling other docker images. Pulling docker images should be a write-intensive task, and I see from `top` that several `[z_wr_iss]` threads were taking a considerable amount of CPU time for several seconds.
2. In another eariler boot I have manually trimmed on the zfs pool - the trim should effectively goes inside the LVM layer; however I am sure that `/usr/bin/find` and `/usr/lib/libpython3.10.so.1.0` are installed after that.

Basically, I guess that using docker and pull several big images while building another image, especially with trimming, may reproduce such problem.

### Describe how to reproduce the problem

Not sure yet. Will add more when I can reliably reproduce the problem and reduce related options to a minimum.

### Include any warning/errors/backtraces from the system logs

n/a

**Comment by TheUbuntuGuy** [comment]
Are the files in the dataset permanently corrupt, or is the corruption transient/in-memory only? i.e. if the system is restarted for example, are the files still zeroed?

**Comment by logoerthiner1** [comment]
> Are the files in the dataset permanently corrupt, or is the corruption transient/in-memory only? i.e. if the system is restarted for example, are the files still zeroed?

Permanent. I have just powered on that VM and inspect the same docker container and `/usr/bin/find` still have same problem.

**Comment by logoerthiner1** [comment]
I have observed the same scenario the second time, this time without trimming. This time the volume is nearly full. Therefore I believe that it can be reproduced when zfs is built on smaller volume and frequent creating file and deleting files (so that most chunks are initialized).

**Comment by logoerthiner1** [comment]
proof of concept

```
FROM alpine:3.16.6

RUN apk update \
    && date \
    && apk upgrade \
    && apk add \
        alpine-sdk bash clang compiler-rt \
        coreutils diffutils elfutils-dev findutils llvm python3 \
    && date \
    && busybox du /usr/bin/coreutils \
    && busybox sha256sum /usr/bin/coreutils && busybox hexdump -C -n 1024 /usr/bin/coreutils \
    && busybox du /usr/bin/coreutils
# correct
# ff9a44075b6d2b17a1cef5fee14084771e0fd981ddb928c4fa8808ceb37dfb60  /usr/bin/coreutils
# wrong

WORKDIR /root
ARG A=1
WORKDIR /
```

The output for coreutils part inside docker build output is:
```
1	/usr/bin/coreutils
ff9a44075b6d2b17a1cef5fee14084771e0fd981ddb928c4fa8808ceb37dfb60  /usr/bin/coreutils
```

the file hash is correct but the file size is incorrect.

however the actual file size as can be seen in the final image is not 1, but rather:
```
bash-5.1# busybox du /usr/bin/coreutils
581	/usr/bin/coreutils
bash-5.1# ls -als /usr/bin/coreutils
581 -rwxr-xr-x 1 root root 1123280 Apr 19  2022 /usr/bin/coreutils
bash-5.1# 
```

Finding out what mechanism caused the file to be of size 1 but full of data is the first step; coreutils is merely one convenient binary that is not a symlink, other binary generated from `apk add` show similar phenomenon

I suspect that zfs implementation can have this time window that a newly created file or a newly cloned file shows blank initially which causes docker to get confused. Or merely docker does clone-copy-clone too fast that the file are copied when it shows zero.

Again this is not the actual bug triggering poc but merely a suspicious phenomenon. Bug is not easy to trigger and may be related to timing and concurrent actions on zfs (for example building two docker images at the same time?).

Any good ideas for diagnosing the problem? or any ideas about my config is wrong?

**Comment by snajpa** [comment]
AFAIK there are some patches related to mmaped files that didn't make it to 2.1x; could you verify if this problem still exists in 2.2.7?
